onnxruntime/core/optimizer/gelu_fusion.cc

```diff
@@ -0,0 +1,204 @@
+    }
+  }
+  return false;
+}
+
+// Gelu supports limited data types.
+static std::vector<std::string> supported_data_types{"tensor(float16)", "tensor(float)", "tensor(double)"};
+
+static bool IsSupportedDataType(const Node& node) {
+  for (const auto& input_arg : node.InputDefs()) {
+    if (std::find(supported_data_types.begin(), supported_data_types.end(),
+                  *(input_arg->Type())) == supported_data_types.end()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+Status GeluFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  std::deque<onnxruntime::NodeIndex> removed_nodes;
+  std::deque<NodeArg*> removed_initializers;
+
+  for (auto node_index : node_topology_list) {
+    auto& div = *graph.GetNode(node_index);
+    ORT_RETURN_IF_ERROR(Recurse(div, modified, graph_level));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(div, "Div", {7}) ||
+        !graph_utils::IsSupportedProvider(div, GetCompatibleExecutionProviders()) ||
+        div.GetOutputEdgesCount() != 1 ||
+        !IsSupportedDataType(div)) {
+      continue;
+    }
+
+    // Check second input is sqrt(2)
+    if (!CheckConstantInput(graph, *(div.MutableInputDefs()[1]), static_cast<float>(M_SQRT2))) {
+      continue;
+    }
+
+    const Node& erf_node = *(div.OutputNodesBegin());
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(erf_node, "Erf", {9}) ||
+        erf_node.GetExecutionProviderType() != div.GetExecutionProviderType() ||
+        erf_node.GetOutputEdgesCount() != 1 ||
+        !IsSupportedDataType(erf_node)) {
+      continue;
+    }
+
+    const Node& add_node = *(erf_node.OutputNodesBegin());
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(add_node, "Add", {7}) ||
+        add_node.GetExecutionProviderType() != div.GetExecutionProviderType() ||
+        add_node.GetOutputEdgesCount() != 1 ||
+        !IsSupportedDataType(add_node)) {
+      continue;
+    }
+
+    // Check the other input node(e.g. not of type Erf) is 1.0f.
+    const Node& add_first_input_node = *(add_node.InputNodesBegin());
+    int add_const_input_index = 0;
+    if (add_first_input_node.OpType().compare("Erf") == 0) {
+      add_const_input_index = 1;
+    }
+    const auto& add_const_input_arg = add_node.InputDefs()[add_const_input_index];
+    if (!CheckConstantInput(graph, *add_const_input_arg, 1.0f)) {
+      continue;
+    }
+
+    const Node& mul_node = *(add_node.OutputNodesBegin());
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(mul_node, "Mul", {7}) ||
+        mul_node.GetExecutionProviderType() != div.GetExecutionProviderType() ||
+        !IsSupportedDataType(mul_node)) {
+      continue;
+    }
+
+    const Node* mul2_node = nullptr;
+    for (auto iter = mul_node.InputNodesBegin(); iter != mul_node.InputNodesEnd(); ++iter) {
+      if ((*iter).OpType().compare("Mul") == 0) {
```

Comment:
Could you use the same approach as with add_node calculating add_const_input_index?

Reply:
This is a bit different logic, add_const_input_index is used to get input NodeArg, but here we want to get input Node. My understanding is we cannot index input node with input index, right?

Reply:
Got it.

---
In reply to: [330875094](https://github.com/microsoft/onnxruntime/pull/1746#discussion_r330875094) [](ancestors = 330875094)